### ISSUE TITLE
Quantization option does not apply.

### ISSUE THREAD

**Issue by miziworld** [espnet/espnet#4041]
Hi, 
As I asked last time I am using the egs/ksponspeech/asr example. As you pointed out, I'm using the quantization option implemented in espnet1's asr_recog.py. However, what I applied does not seem to be reflected during actual decoding.
I used the quantization dtype as qint8 and gave the quantize-asr-model true option, but the decoding time was the same as without the quantization option.
Is there something I'm using wrong?
Also, if quantization is applied, how can we numerically check the applied weight value and the reduced amount of computation?
`        asr_recog.py \                                                                                                                                                              --quantize-dtype qint8 \                                                                                                                                                    --quantize-config Linear \                                                                                                                                                  --quantize-asr-model true \                                                                                                                                                 --config ${decode_config} \                                                                                                                                                 --ngpu ${ngpu} \                                                                                                                                                            --backend ${backend} \                                                                                                                                                      --debugmode ${debugmode} \                                                                                                                                                  --verbose ${verbose} \                                                                                                                                                      --recog-json ${feat_recog_dir}/data.json \                                                                                                                                  --result-label ${decode_dir}/result.json \                                                                                                                                  --model ${recog_model} \                                                                                                                                                    --api ${api} \                                                                                                                                                              ${recog_opts}        `

and this is the result of decode.log
![image](https://user-images.githubusercontent.com/17849068/152769116-c9e61062-3014-4824-9583-44cf507ddd17.png)

**Comment by b-flo** [comment]
Hi,

The logging message doesn't take into account quantization dtype (`args.quantize_dtype`) it seems and initial dtype (`args.dtype`) is also not changed:

```
    dtype = getattr(torch, args.dtype)
    logging.info(f"Decoding device={device}, dtype={dtype}")
```

> Also, if quantization is applied, how can we numerically check the applied weight value and the reduced amount of computation?

By checking/comparing model parameters. Not sure if we have the utilities in ESPnet1 but you can take [this](https://github.com/espnet/espnet/blob/master/espnet2/torch_utils/model_summary.py) as example to do it yourself.

**Comment by b-flo** [comment]
> If quantization is properly applied, why is there no difference in decoding speed and results before and after application?

Sorry, I was only pointing out that the message you highlighted displays the wrong information! Not sure if everything is working properly, could you check the model parameters?

> Also, in espnet1 now, I know that quantization is possible only in inference, but if it is possible to change the dtype during training, it will be much more efficient. Is such a method implemented?

Quantization aware training is not supported currently. It's on the TODO list of multiple ESPnet members but it's not a priority (at least for me), so I can't really comment on an ETA..

**Comment by miziworld** [comment]
I tried to run model_summary.py in espnet1 with the method you pointed out, but in espnet1 the model doesn't seem to be saved as pth. Could you please help me how to use model_summary.py?

**Comment by miziworld** [comment]
Sorry for asking a basic question. 
I am using pytorch as a backend, and I understand that to call the class you pointed out, load the model like torch.load(PATH+'model.pt') and then pass it as an argument.
In recog_wav.sh, asr_recog.py is called and the model is loaded at this time. I need help on where the model to which the quantization option is applied is stored and how to use model_summary().

**Comment by b-flo** [comment]
Looking at your decoding parameters, you're using [recog_v2](https://github.com/espnet/espnet/blob/master/espnet/asr/pytorch_backend/recog.py#L23). In this method, the original model is loaded [here](https://github.com/espnet/espnet/blob/master/espnet/asr/pytorch_backend/recog.py#L44) and quantized [here](https://github.com/espnet/espnet/blob/master/espnet/asr/pytorch_backend/recog.py#L75). After each call, you can write `logging.info(model_summary(model))`

> the quantization option is applied is stored ...

Hum, I'm not sure we're saving/writing the quantized model when using dynamic quantization. Let me check.

**Comment by miziworld** [comment]
Thank you so much for the great help!!
As a result of checking, the number of parameters and the model size seem to be properly quantized, but why is there no difference in decoding speed?

I just have one more question to ask.
Currently, only one wav file is tested using recog_wav.sh, but is there a way to put the test wav files without a script in the folder and run recog_wav.sh to decode them at once? (For example, recog_wav. A method of decoding at the 3rd decoding stage sequentially after processing at once in stages 1 and 2 of sh)

**Comment by b-flo** [comment]
> but why is there no difference in decoding speed?

Hum, I'm not sure about that. It can be explained by the model architecture and what could be quantized, or how it would impact the search process for example.

> Currently, only one wav file is tested using recog_wav.sh, but is there a way to put the test wav files without a script in the folder and run recog_wav.sh to decode them at once? (For example, recog_wav. A method of decoding at the 3rd decoding stage sequentially after processing at once in stages 1 and 2 of sh)

I'm not sure I totally understood but if you're asking if it's possible to use a script to decode, from anywhere, a batch of wav files instead of a single wav file, the answer is yes. However, I'm not sure we have a script available for that, you will need to make it yourself (or modify `recog_wav.sh`).
If you're a bit familiar with bash and ESPnet (at least the Kaldi-like data preparation), it won't take long depending on your solution.
